feat(website): add empty state pattern docs (#1371)
twilio-labs/paste: packages/paste-website/src/pages/patterns/empty-states/index.mdx

import {Table, Tr, Th, Td, THead, TBody} from '@twilio-paste/table';
import {Text} from '@twilio-paste/text';
import {useUID} from 'react-uid';

export const pageQuery = graphql`
  {
    mdx(fields: {slug: {eq: "/patterns/empty-states/"}}) {
      fileAbsolutePath
      frontmatter {
        slug
        title
      }
      headings {
        depth
        value
      }
    }
  }
`;

<content>

<Breadcrumb>
  <BreadcrumbItem to="/">Home</BreadcrumbItem>
  <BreadcrumbItem to="/patterns">Patterns</BreadcrumbItem>
</Breadcrumb>
<Heading as="h1" variant="heading10">
  {props.pageContext.frontmatter.title}
</Heading>
<Paragraph>{props.pageContext.frontmatter.description}</Paragraph>

<Box as="dl" display="flex">
  <Text as="dt" color="colorTextWeak" lineHeight="lineHeight30">
    <Box width="size10">Status</Box>
  </Text>
  <Text as="dd">beta</Text>
</Box>

</content>

---

<contentwrapper>

<PageAside data={props.data.mdx} />

<content>

## Ingredients

<Grid gutter="space30" marginBottom="space80">
  <Column>
    <Card>
      <Heading as="h3" variant="heading40" marginBottom="space0">
        <Anchor href="../components/card">Card</Anchor>

>       </Heading>

Reviewer: Can you add a paragraph to this section that says

> Go to the [Sizing & responsiveness section](link to sizing & responsiveness) for more details.

Author: @serifluous Does this work? https://share.getcloudapp.com/4gu2Ze8B

I tried adding the copy to the Card and Illustration boxes, but that really throws off the box layout.

    </Card>
  </Column>
  <Column>
    <Card>
      <Heading as="h3" variant="heading40" marginBottom="space0">
        <Anchor href="../components/heading">Heading</Anchor>
      </Heading>
    </Card>
  </Column>
  <Column>
    <Card>
      <Heading as="h3" variant="heading40" marginBottom="space0">
        <Anchor href="../components/paragraph">Paragraph</Anchor>
      </Heading>
    </Card>
  </Column>
</Grid>
<Grid gutter="space30" marginBottom="space80">
  <Column>
    <Card>
      <Heading as="h3" variant="heading40" marginBottom="space0">
        <Anchor href="../components/button">Button (optional)</Anchor>
      </Heading>
    </Card>
  </Column>
  <Column>
    <Card>
      <Heading as="h3" variant="heading40" marginBottom="space0">
        <Anchor href="../components/anchor">Anchor (optional)</Anchor>
      </Heading>
    </Card>
  </Column>
  <Column>
    <Card>
      <Heading as="h3" variant="heading40" marginBottom="space0">
        <Anchor href="../illustrations">Illustrations (optional)</Anchor>

>       </Heading>

Reviewer: Can you add paragraphs to this card that say:

> We recommend using a fixed width of `$size-20`.
> 
> The illustration must be directly related to the message you’re communicating. Read more about [which illustrations to use for each use case](link to "Content examples by use case").

    </Card>
  </Column>
</Grid>

```jsx
// import all ingredients for the empty state patterns
import {​Card} from '@twilio-paste/core/card';
import {Heading} from '@twilio-paste/core/heading';
import {Paragraph} from '@twilio-paste/core/paragraph';
import {Button} from '@twilio-paste/core/button';
import {​Anchor} from '@twilio-paste/core/anchor';
```

## Usage

### General

This pattern is used to explain to a customer what they should do next if content is either unavailable or can't be displayed.

Empty states should:

- **Be clear:** If there's a path forward, explain it as concisely as possible. Include an explanation for why a customer is seeing the empty state if it helps them better understand what to do next.

> - **Give direction and educate the customer:** Let customers know what steps they can take to move forward in a flow. Guiding them to "Create an alpha sender" is more helpful than saying "No alpha senders".

Reviewer 2: What does "in a flow" mean when we say "move forward in a flow"? I was confused by it and would suggest removing it and having the sentence read "Let customers know what steps they can take to move forward."

Author: I think it's referring to a user flow. I like your suggestion. Thanks!

- **Be encouraging and inspire confidence:** Use a [friendly and approachable tone](/content/voice-and-tone#encouraging) to ensure a positive experience. Never make customers feel unsuccessful or guilty because they're seeing an empty state. Even when they arrive at a page "in error" (for example, on a page they don't have permissions to), encouraging messaging and an explanation of possible solutions can give customers confidence that they're heading in the right direction.

In most cases, treat an empty state as an informational state of the entire page, and not a state that simply replaces a part of the page that should otherwise be populated. When you have text explaining what a page is for, try working it into the empty state card.

> When a page is in an empty state, its call-to-action (CTA) should appear in the empty state card, not near the page heading like in the [Create pattern](/patterns/create/). Exceptions can be made in cases like some "No results found" scenarios, when a CTA is already on the page on load. The visually focus of hiding the CTA that's outside of the card could draw too much of the customer's attention away from their primary goal.

Reviewer 2: There's a lot of good info and recommendations here! I cleaned it so that there's more clarity in the guidance. Lemme know what you think:

Treat an empty state as an informational state of the _entire page_, not a replacement of the part of the page that should otherwise be populated. When a page is in an empty state, the call-to-action (CTA) should appear in the empty state card, not near the page heading like in the [Create pattern](/patterns/create/).

Exceptions can be made in cases where a CTA is already on the page on load- for example, in some "No results found" scenarios.

Reviewer: I'm worried that folks will take the "Only one CTA on the page" guideline to the extreme, where if there's already a button on the top right on page load, then they'll hide it in order to show the button in the empty state (in a "no results found" case, for example).

It's a little hard to explain through words which might be why the text here isn't super clear. I can maybe show what I'm trying to say in standup? Def open to improvements on wording.

> When a page is in an empty state, its call-to-action (CTA) should appear in the empty state card, not near the page heading like in the [Create pattern](/patterns/create/). Exceptions can be made in cases like some "No results found" scenarios, when a CTA is already on the page on load. The visually focus of hiding the CTA that's outside of the card could draw too much of the customer's attention away from their primary goal.

Reviewer: ```suggestion
When a page is in an empty state, its call-to-action (CTA) should appear in the empty state card, not near the page heading like in the [Create pattern](/patterns/create/). Exceptions can be made in cases like some "No results found" scenarios, when a CTA is already on the page on load. The visual focus of hiding the CTA that's outside of the card could draw too much of the customer's attention away from their primary goal.
```

> <Box marginBottom="space70">

Reviewer 2: I'm really liking these code sandboxes included w/in the page. I think it makes it easy to see how the pattern is created and will be super useful!!

  <iframe
    src="https://codesandbox.io/embed/empty-state-with-immediate-next-step-bmxe1?fontsize=14&hidenavigation=1&theme=dark&view=preview"
    style="width:100%; height:350px; border:0; border-radius: 4px; overflow:hidden;"
    title="Empty state requiring a customer to wait"
    allow="accelerometer; ambient-light-sensor; camera; encrypted-media; geolocation; gyroscope; hid; microphone; midi; payment; usb; vr; xr-spatial-tracking"
    sandbox="allow-forms allow-modals allow-popups allow-presentation allow-same-origin allow-scripts"
  ></iframe>

> </Box>

Reviewer: Did using SVGs not work? Should we include a callout that tells people to use PNGs for now?

Author: We can update the codesandboxes with the svgs. I used pngs to just get something up there for now.

Author: @serifluous I took at this, and we'll need to do some more work to get svgs working in our codesandboxes. The svgs are also HUGE: ~400kb vs ~60kb for the pngs. I think we should leave them as pngs for now. Downstream teams will most likely have a way to handle svg loading already.

Reviewer 3: Until we have themeable illustrations I think PNGs are fine.

### Accessibility

For call-to-actions in an empty state, use an [Anchor](/components/anchor) if the action will take the user to a new URL.
In all other cases, use a [Button](/components/button). For detailed guidance on applying this recommendation,

> check out the [Button vs. Anchor pattern](/patterns/button-vs-anchor).

Reviewer 2: YES button vs. anchor page coming in REAL handy 🎉

    title="Empty state requiring a customer to wait"
    allow="accelerometer; ambient-light-sensor; camera; encrypted-media; geolocation; gyroscope; hid; microphone; midi; payment; usb; vr; xr-spatial-tracking"
    sandbox="allow-forms allow-modals allow-popups allow-presentation allow-same-origin allow-scripts"
  ></iframe>
</Box>

### Sizing & Responsiveness

When an empty state card is wider than 540px (recommended), use `padding="space150"`, and align illustrations horizontally with the text content.

In smaller spaces, use `padding="space100"` and stack the illustration in the card above the text content.

If your empty state composition needs to accommodate different screen sizes, you can also wrap it in the Stack or Grid component.

<Callout>
  <CalloutText>
    If you'd like to contribute guidelines for empty states in even more compact situations, like a navigation panel or
    components like <Anchor href="/components/combobox">Combobox</Anchor>, please reach out in a{' '}
    <Anchor href="https://github.com/twilio-labs/paste/discussions">GitHub Discussion</Anchor>!
  </CalloutText>
</Callout>

> 

Reviewer: ```suggestion

Open the sandbox in a new tab and resize the browser for an example of responsive sizing.

```

    title="Empty state requiring a customer to wait"
    allow="accelerometer; ambient-light-sensor; camera; encrypted-media; geolocation; gyroscope; hid; microphone; midi; payment; usb; vr; xr-spatial-tracking"
    sandbox="allow-forms allow-modals allow-popups allow-presentation allow-same-origin allow-scripts"
  ></iframe>
</Box>

> ## Content examples by use case

Reviewer 2: This heading was hard for me to understand. Maybe we could switch it to something like "Examples of Empty States"?

Reviewer: hm, true. Let's go with "Examples of Empty State use cases"

> Use a heading that succinctly summarizes the situation. We recommended a maximum of 7 words.

Reviewer: ```suggestion
Use a heading that succinctly summarizes the situation. We recommend a maximum of 7 words.
```

>   ></iframe>

Reviewer: Can either you or @TheSisb update this sandbox so the bottom 2 cards are side by side in the same row in all cases? It's a little hard to tell how it's resizing without opening the sandbox in a new tab, and then the sandbox preview again in a new tab.

Author: Yeah for sure. Done!

Use an illustration that's directly related to the message you're communicating.

> Don't use a random illustration to "fill in the blank." The more that illustrations are used throughout our products, the more they lose their emotional and visual impact when it really matters.

Reviewer 2: Should these recs live elsewhere? They don't match the heading above them. Maybe the heading should be changed to include them and say "Building an Empty State" and then have another subheading above the table that says "Use Cases"

Reviewer: Would changing it to this help?

> For all use cases:
> - Use a heading that succinctly summarizes the situation. We recommended a maximum of 7 words.
> - Use an illustration only if it’s directly related to the message you’re communicating. Don’t use a random illustration to “fill in the blank.” The more that illustrations are used throughout our products, the more they lose their emotional and visual impact when it really matters. Learn more about requesting a new illustration.

Eventually we'll want to rethink IA for these Patterns pages because we have composition guidance both at the top of the page and down below.

> Use a heading that succinctly summarizes the situation. We recommend a maximum of 7 words.
> 
> Use an illustration that's directly related to the message you're communicating.
> 
> Don't use a random illustration to "fill in the blank." The more that illustrations are used throughout our products, the more they lose their emotional and visual impact when it really matters.

Reviewer: ```suggestion
For all use cases:

- Use a heading that succinctly summarizes the situation. We recommended a maximum of 7 words.
- Use an illustration only if it’s directly related to the message you’re communicating. Don’t use a random illustration to “fill in the blank.” The more that illustrations are used throughout our products, the more they lose their emotional and visual impact when it really matters. Learn more about requesting a new illustration.
```

> Use a heading that succinctly summarizes the situation. We recommend a maximum of 7 words.
> 
> Use an illustration that's directly related to the message you're communicating.
> 
> Don't use a random illustration to "fill in the blank." The more that illustrations are used throughout our products, the more they lose their emotional and visual impact when it really matters.
> 
> [Learn more about requesting a new illustration](/illustrations#request-new-illustrations).

Reviewer: ```suggestion
For all use cases:

- Use a heading that succinctly summarizes the situation. We recommended a maximum of 7 words.
- Use an illustration only if it’s directly related to the message you’re communicating. Don’t use a random illustration to “fill in the blank.” The more that illustrations are used throughout our products, the more they lose their emotional and visual impact when it really matters. [Learn more about requesting a new illustration](/illustrations#request-new-illustrations).

```

            maxWidth="100px"
            height="auto"
          />
          <Anchor href="https://library.twilio.com/collections/view/963D9AB3-47A0-4C13-978BC97108C1EBCC/?mediaId=8C1B7853-CB10-4C2E-89BF03871593BDC3">
            Download from the Brand library
          </Anchor>
        </Td>
      </Tr>
      <Tr verticalAlign="top">
        <Th scope="row">Change regional data location</Th>
        <Td>The customer needs to swap regions to access the data.</Td>
        <Td>"View event logs in the United States region"</Td>
        <Td>
          <Box
            as="img"
            src="/images/patterns/empty-change-regional-data-location.png"
            alt="Change regional data location illustration"
            display="block"
            marginLeft="auto"
            marginRight="auto"
            maxWidth="100px"
            height="auto"
          />
          <Anchor href="https://library.twilio.com/collections/view/963D9AB3-47A0-4C13-978BC97108C1EBCC/?mediaId=F44538E4-FD2E-45B2-806402557124153C">
            Download from the Brand library
          </Anchor>
        </Td>
      </Tr>
    </TBody>
  </Table>
</Box>

## Dos and Don'ts

<DoDont>
  <Do
    title="Do"

>     body="Include a heading that summarizes the situation. We recommended a maximum of 7 words."

Reviewer: ```suggestion
    body="Include a heading that summarizes the situation. We recommend a maximum of 7 words."
```

    preview={false}
  />
  <Dont
    title="Don't"

>     body="Don't include a long heading providing too many details. If the heading wraps to two or more lines, that may be a sign to be more concise. Don't use a period at the end of a heading."

Reviewer: ```suggestion
    body="Don't include a long heading providing too many details. If the heading wraps to 2 or more lines, that may be a sign to be more concise. Don't use a period at the end of a heading."
```

    preview={false}
  />
  <Dont title="Don't" body="Avoid showing the same CTA twice on a page." preview={false} />
</DoDont>

<DoDont>
  <Do title="Do" body="If the empty state requires CTAs, try to include no more than 2." preview={false} />
  <Dont
    title="Don't"
    body="Avoid giving the user more than 3 calls-to-action on the same empty state, so they're not overwhelmed with choice. If more than 3 are needed, consider grouping them under different headings or other content sections."
    preview={false}
  />
</DoDont>

<DoDont>
  <Do
    title="Do"
    body="Use an illustration that's directly related to the message you are sending. Check the list of illustrations to use based on use case."
    preview={false}
  />
  <Dont
    title="Don't"
    body='Don&apos;t use a random illustration to "fill in the blank." The more that illustrations are used throughout our products, the more they lose their emotional and visual impact when it really matters.'
    preview={false}
  />
</DoDont>

## Starter kits

> ### CodeSandbox
> 
> Coming soon
> 

Reviewer: ```suggestion
```

Reviewer: We can remove this since there are CodeSandbox examples on the page.
